File: 백준/Silver/15663. N과 M （9）/N과 M （9）.py
n, m = map(int, input().split())
arr = list(map(int, input().split()))

arr.sort()
visited = [0] * n

def dfs(depth):
    if depth == m:
        temp = " ".join(map(str, box))
        if temp not in answer:
            answer.append(temp)
        return

    for i in range(n):
        if visited[i] == 0:
            visited[i] = 1
            box.append(arr[i])
            dfs(depth + 1)
            box.pop()
            visited[i] = 0

box = []
answer = []
dfs(0)

print("\n".join(answer))

File: 백준/Silver/15663. N과 M （9）/N과 M （9）.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "ex.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const [n, m] = input[0].split(" ").map(Number);
const arr = input[1].split(" ").map(Number);
arr.sort((a, b) => a - b);

const result = [];
const answer = new Set();
const visited = Array(n).fill(0);

const dfs = (depth) => {
  if (depth === m) {
    answer.add(result.join(" "));
    return;
  }

  for (let i = 0; i < arr.length; i++) {
    if (visited[i]) continue;
    result.push(arr[i]);
    visited[i] = 1;
    dfs(depth + 1);
    visited[i] = 0;
    result.pop();
  }
};

dfs(0);

answer.forEach((a) => {
  console.log(a);
});
